src/crypto/tls/ech_provider.go

```diff
@@ -57,22 +68,239 @@ type ECHProviderResult struct {
 	// by the provider and no error was reported. The data has the following
 	// format (in TLS syntax):
 	//
-	// enum { sender(0), receiver(1) } HpkeRole;
+	// enum { sealer(0), opener(1) } HpkeRole;
 	//
 	// struct {
 	//     HpkeRole role;
-	//     HpkeKemId kem_id;   // as defined in draft-irtf-cfrg-hpke-05
-	//     HpkeKdfId kdf_id;   // as defined in draft-irtf-cfrg-hpke-05
-	//     HpkeAeadId aead_id; // as defined in draft-irtf-cfrg-hpke-05
+	//     HpkeKemId kem_id;   // as defined in draft-irtf-cfrg-hpke-06
+	//     HpkeKdfId kdf_id;   // as defined in draft-irtf-cfrg-hpke-06
+	//     HpkeAeadId aead_id; // as defined in draft-irtf-cfrg-hpke-06
 	//     opaque exporter_secret<0..255>;
 	//     opaque key<0..255>;
-	//     opaque nonce<0..255>;
-	//     uint64 seq;
+	//     opaque base_nonce<0..255>;
+	//     opaque seq<0..255>;
 	// } HpkeContext;
-	//
-	// NOTE(cjpatton): This format is specified neither in the ECH spec nor the
-	// HPKE spec. It is the format chosen for the HPKE implementation that we're
-	// using. See
-	// https://github.com/cisco/go-hpke/blob/9e7d3e90b7c3a5b08f3099c49520c587568c77d6/hpke.go#L198
 	Context []byte
 }
+
+// EXP_ECHKeySet implements the ECHProvider interface for a sequence of ECH keys.
+type EXP_ECHKeySet struct {
```

Comment:
The temptation is to go ahead and update while this is in use, with consequent terrible problems. Either this should be documented as something you shouldn't do, or there should be locks.

Reply:
> The temptation is to go ahead and update while this is in use, with consequent terrible problems.

I don't know what you mean here. Can you elaborate?

Added the following line to the comment: "NOTE: This API is EXPERIMENTAL and subject to change."

Reply:
so EXP_ECHKeySet is something a user would  pass into a tls configuration. But then the user wants to update it to implement key rotation, and it isn't threadsafe.

Reply:
I see. EXP_ECHKeySet is only meant to provide basic functionality. In particular, the set of supported ECH keys doesn't change over the lifetime of the TLS config. Should I document this?